[quant][embedding qat] eager mode QAT for Embeddings

b-koopman opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68390
* __->__ #66429

Summary:

Add QAT for Embeddings via eager mode API.

To avoid backward compatibility issues, this has not yet been added to the default QAT mappings.
For examples of use, see tests in "Test Plan" below.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "test_embedding_qat_config"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "test_embedding_qat_qconfig_equal" `

Reviewers:
supriyar, HDCharles

Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Move module mappings out of default

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

b-koopman committed:
[quant][embedding qat] eager mode QAT for Embeddings

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

supriyar commented on test/quantization/eager/test_quantize_eager_qat.py:
```diff
@@ -109,17 +119,43 @@ def checkQuantized(model):
+
+                test_only_train_fn(model, self.embed_linear_data_train)
+                # make activation_post_process is not inserted for Embedding
+                self.assertFalse(hasattr(model, "activation_post_process"))
```
Shouldn't an activation_post_process be inserted for linear layers here?

supriyar commented on torch/ao/quantization/quantization_mappings.py:
```diff
@@ -44,6 +44,7 @@
     nn.ELU: nnq.ELU,
     nn.Embedding: nnq.Embedding,
     nn.EmbeddingBag: nnq.EmbeddingBag,
+    nn.Embedding: nnq.Embedding,
```
This is the same as line 45

supriyar commented on torch/nn/qat/modules/embedding_ops.py:
```diff
@@ -3,6 +3,72 @@
+                                           self.sparse, None, self.device, self.dtype)
+        embedding_bag.weight = torch.nn.Parameter(self.weight.detach())
+        embedding_bag.train(self.training)
+        return embedding_bag
```
nit: add a new line after this

supriyar commented on torch/nn/qat/modules/embedding_ops.py:
```diff
@@ -61,7 +127,6 @@ def from_float(cls, mod):
         qat_embedding_bag = cls(mod.num_embeddings, mod.embedding_dim, mod.max_norm, mod.norm_type,
                                 mod.scale_grad_by_freq, mod.mode, mod.sparse, mod.weight,
                                 mod.include_last_offset, mod.padding_idx, qconfig=qconfig)
-        qat_embedding_bag.weight = mod.weight
```
why did we remove this line?

b-koopman replied:
This is not required as mod.weight is being passed as a parameter to the constructor.

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"




Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Summary:

Add QAT for Embeddings via eager mode API.

To avoid backward compatibility issues, this has not yet been added to the default QAT mappings.
For examples of use, see tests in "Test Plan" below.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "test_embedding_qat_config"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "test_embedding_qat_qconfig_equal" `

Reviewers:
supriyar, HDCharles

Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Move module mappings out of default

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

b-koopman committed:
Update on "[quant][embedding qat] eager mode QAT for Embeddings"


Summary:

Add QAT for Embeddings via eager mode API.

To avoid backward compatibility issues, this has not yet been added to the default QAT mappings.
For examples of use, see tests in "Test Plan" below.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "test_embedding_qat_config"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "test_embedding_qat_qconfig_equal" `

Reviewers:
supriyar, HDCharles

Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Move module mappings out of default

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).
